middle_end/flambda2/tests/lib/minicheck/type.ml

```diff
@@ -0,0 +1,270 @@
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*           Mark Shinwell and Leo White, Jane Street Europe              *)
+(*                                                                        *)
+(*   Copyright 2015--2022 Jane Street Group LLC                           *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+[@@@ocaml.warning "+a-4-30-40-41-42"]
+
+type 'a t =
+  { generate : Splittable_random.t -> 'a;
+    print : Format.formatter -> 'a -> unit
+  }
+
+let generate t = t.generate
+
+let print t = t.print
+
+let print_unknown ppf _ = Format.pp_print_string ppf "_"
+
+let define ~generate ?(print = print_unknown) () = { generate; print }
+
+let with_print t ~print = { t with print }
+
+let bool =
+  let generate r = Splittable_random.int r mod 2 = 0 in
+
+  define () ~generate ~print:Format.pp_print_bool
+
+let int = define () ~generate:Splittable_random.int ~print:Format.pp_print_int
+
+let map_generate ty ~f =
+  let generate r = f (generate ty r) in
+  define () ~generate ~print:(print ty)
+
+(* Should be small because we're doing this the dumb way *)
+let small_nat ~less_than =
+  map_generate int ~f:(fun i -> Int.abs i mod less_than)
+
+let log_int =
+  let generate r =
+    (* This assumes that the size of [int] is one less than the word size, which
+       isn't true for JavaScript *)
+    let bits = Splittable_random.int r mod Sys.word_size in
+    if bits = 0
+    then 0
+    else
+      let high_bit = 1 lsl (bits - 1) in
+      let mask = high_bit - 1 in
+      Splittable_random.int r land mask lor high_bit
+  in
+  define () ~generate ~print:Format.pp_print_int
+
+let option ty =
+  let generate r =
+    let choose_none = generate bool r in
+    if choose_none then None else Some (generate ty r)
+  in
+  let print ppf o =
+    Format.pp_print_option
+      ~none:(fun ppf () -> Format.pp_print_string ppf "<none>")
+      (print ty) ppf o
+  in
+  define () ~generate ~print
+
+let map { generate; print = _ } ~f =
+  let generate r = f (generate r) in
+  define () ~generate
+
+let bind { generate; print = _ } ~f =
+  let generate r =
+    let a = generate r in
+    let { generate; print = _ } = f a in
+    generate r
+  in
+  define () ~generate
+
+(* Would love to make these print functions local, but doing so crashes
+   ocamlformat due to the [@ocamlformat "disable"]. See
+   https://github.com/ocaml-ppx/ocamlformat/issues/2096 *)
```

Comment:
Oh, we do that with all the `print` functions in Flambda2 because we want to use `%a` formatters and group together the function-argument pairs:

```ocaml
Format.fprintf ppf "%a %a"
  print_foo f
  print_bar b
```
In longer examples, `ocamlformat` will often put line breaks in awkward places like between `print_barl` and `b`

Reply:
Oh, excellent!  That's valuable information, thanks